core/src/main/java/org/infinispan/persistence/async/AsyncCacheWriter.java

```diff
@@ -123,12 +116,21 @@ public Thread newThread(Runnable r) {
    @Override
    public void stop() {
       if (trace) log.tracef("Stop async store %s", this);
-      stateLock.writeLock(1);
-      state.get().stopped = true;
+      stateLock.writeLock(0);
+      stopped = true;
       stateLock.writeUnlock();
       try {
-         coordinator.join(shutdownTimeout);
-         if (coordinator.isAlive())
+         // It is safe to wait without timeout because the thread pool uses an unbounded work queue (i.e.
+         // all work handed to the pool will be accepted and eventually executed) and AsyncStoreProcessors
```

Comment:
The fact that the thread pool uses an unbounded queue doesn't say too much about being to finish in a bounded amount of time. I think it would be more relevant to put this comment where the executor is created, to warn that a bounded queue with a discard/abort policy would lose data.